File: src/scitrader.marketdataserver/src/main/java/com/scitrader/marketdataserver/exchange/bitmex/BitmexWebsocketClient.java
package com.scitrader.marketdataserver.exchange.bitmex;


import com.jsoniter.JsonIterator;
import com.jsoniter.any.Any;
import com.scitrader.marketdataserver.common.MarketDataServerException;
import com.scitrader.marketdataserver.transport.SciTraderWebsocketClient;
import com.sun.javaws.exceptions.InvalidArgumentException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;

import java.net.URI;
import java.net.URISyntaxException;

public class BitmexWebsocketClient implements IBitmexWebsocketClient{

  Logger Log = LogManager.getLogger(BitmexWebsocketClient.class);

  private final String apiUrl = "wss://www.bitmex.com/realtime";
  private final String[] instruments = new String[] { "XBTUSD" };

  public BitmexWebsocketClient() {

  }

  public void connect()  {

    URI uri = getUri();
    Log.info("Initializing SciTraderWebsocketClient with Uri = " + uri.toString());

    SciTraderWebsocketClient ws = new SciTraderWebsocketClient(uri,
            o -> Log.info("Opened websocket"),
            m -> onMessage(m),
            (i,s,b) -> Log.info("Socket closed..."),
            ex -> Log.info("Socket error!"));

    Log.info("Starting connection");

    ws.connect();
  }

  private void onMessage(String message) {
    //Any trade = JsonIterator.deserialize(message);

    if (message.contains("info")) {
      InfoMessage info = JsonIterator.deserialize(message, InfoMessage.class);
      Log.info("Received info: " + info.getInfo());

    } else if (message.contains("success")) {
      SubscribeMessage sub = JsonIterator.deserialize(message, SubscribeMessage.class);
      Log.info("Received Subscribe! " + sub.getRequest().getArgs());
    } else if (message.contains("table") && message.contains("trade")) {
      if (message.contains("partial")){
        Log.info("Ignoring partial trade message: " + message);
      }
      else if (message.contains("insert")){
        TicksMessage tick = JsonIterator.deserialize(message, TicksMessage.class);
        Log.info("Received data: " + tick.data);
      }
    }
    else {
      throw new MarketDataServerException("The message type has no associated deserializer. Message=" + message);
    }

  }

  private URI getUri(){
    try{
      StringBuilder sb  = new StringBuilder();
      sb.append(apiUrl);
      //sb.append("?subscribe=instrument");
      sb.append("?subscribe=trade:XBTUSD,trade:ETHUSD");
//      for(String inst : instruments){
//        sb.append(",trade:").append(inst);
//      }
      return new URI(sb.toString());
    }
    catch(Exception ex){
      throw new MarketDataServerException(ex);
    }
  }
}

File: Kotlin/hello-kotlin-world/src/main/java/com/scitrader/marketdataserver/exchange/bitmex/BitmexWebsocketClient.kt
package com.scitrader.marketdataserver.exchange.bitmex

import com.google.inject.Inject
import com.jsoniter.JsonIterator
import com.mongodb.client.MongoCollection
import com.mongodb.client.MongoDatabase
import com.scitrader.marketdataserver.common.MarketDataServerException
import com.scitrader.marketdataserver.datastore.IMongoDbService
import com.scitrader.marketdataserver.exchange.bitmex.messages.InfoMessage
import com.scitrader.marketdataserver.exchange.bitmex.messages.SubscribeMessage
import com.scitrader.marketdataserver.exchange.bitmex.messages.Tick
import com.scitrader.marketdataserver.exchange.bitmex.messages.TicksMessage
import com.scitrader.marketdataserver.transport.AutoReconnectWebsocket
import org.apache.logging.log4j.LogManager
import org.bson.Document
import java.net.URI
import java.util.concurrent.atomic.AtomicLong

class BitmexWebsocketClient : IBitmexWebsocketClient {

    companion object {
        private val Log = LogManager.getLogger(BitmexWebsocketClient::class.java)
        private val apiUrl = "wss://www.bitmex.com/realtime"
    }

    private var mongoDatabase: MongoDatabase
    private var messageCount : AtomicLong = AtomicLong(0)

    @Inject
    constructor(mongoDbService: IMongoDbService) {

        this.mongoDatabase = mongoDbService.tickDatabase
    }

    override fun connect() {

        val uri = getUri()
        Log.info("Initializing AutoReconnectWebsocket with Uri = " + uri.toString())

        val ws = AutoReconnectWebsocket(uri, { m -> onMessage(m)})
        ws.connect()
    }

    private fun onMessage(message: String) {
        if (message.contains("info")) {
            val info = JsonIterator.deserialize(message, InfoMessage::class.java)
            Log.info("Received info: " + info.info)

        } else if (message.contains("success")) {
            val sub = JsonIterator.deserialize(message, SubscribeMessage::class.java)
            Log.info("Received Subscribe! " + sub.request!!.args)
        } else if (message.contains("table") && message.contains("trade")) {
            if (message.contains("partial")) {
                Log.info("Ignoring partial trade message: $message")
            } else if (message.contains("insert")) {
                val ticks = JsonIterator.deserialize(message, TicksMessage::class.java)

                val mc : Long = messageCount.addAndGet(1)
                if (mc % 100 == 0L){
                    Log.info("($mc th Message) Received data: " + ticks.data)
                }

                for (t in ticks.data) {
                    val collection = getMongoCollection(t)
                    val doc = t.toBsonDocument()
                    collection.insertOne(doc)
                }
            }
        } else {
            throw MarketDataServerException("The message type has no associated deserializer. Message=$message")
        }
    }

    private fun getMongoCollection(t: Tick): MongoCollection<Document> {
        val symbol = t.symbol
        return mongoDatabase.getCollection("BITMEX:$symbol")
    }

    private fun getUri(): URI {
        try {
            val sb = StringBuilder()
            sb.append(apiUrl)

            // TODO: Configuration file for instruments, or subscribe to all instruments
            sb.append("?subscribe=trade:XBTUSD,trade:ETHUSD")

            return URI(sb.toString())
        } catch (ex: Exception) {
            throw MarketDataServerException(ex)
        }
    }

}